### PR TITLE
Move ownership of Waveform into Data.

### PR DESCRIPTION
**Pull request by BenjaminRodenberg**
## Main changes of this PR

Move Waveform into Data. Let users of Waveform access pointer to Waveform via Data.

## Motivation and additional information

Discussion [here](https://github.com/precice/precice/pull/1133#discussion_r746783602) and observation [here](https://github.com/BenjaminRodenberg/precice/pull/1/files#r745679418)

## Author's checklist

* [ ] Factor relevant changes out of this PR and move into https://github.com/precice/precice/pull/1133
* [ ] I added a changelog file with this PR number in `docs/changelog/` if there are noteworthy changes.
* [ ] I ran `tools/formatting/check-format` and everything is formatted correctly.
* [ ] I sticked to C++14 features.
* [ ] I sticked to CMake version 3.10.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: src/cplscheme/CouplingData.hpp [hunk 1]
```diff
@@ -45,6 +43,12 @@ class CouplingData {
   ///  True, if the data values of this CouplingData require to be initialized by this participant.
   const bool requiresInitialization;
 
+  void initializeStorage();
+
+  void moveToNextWindow();
+
+  void storeDataInWaveform();
+
```
Comment (uekerman) [inline src/cplscheme/CouplingData.hpp:51]: Wouldn't it be easier to read if the first two functions also somehow have "waveform" in their name? At least the first one, `initializeWaveformStorage`.

Don't forget the documentation :grin:

File: src/mesh/Data.hpp [hunk 2]
```diff
@@ -23,6 +24,9 @@ namespace mesh {
  */
 class Data {
 public:
+  static const int EXTRAPOLATION_ORDER; // @todo currently hard-coded; we don't care about extrapolation here.
+  static const int INTERPOLATION_ORDER; // @todo currently hard-coded; should be configurable.
```
Comment (uekerman) [inline src/mesh/Data.hpp:28]: `static` brings many problems, see https://github.com/precice/precice/issues/385
Let's be careful here.

File: src/precice/tests/SerialTests.cpp [hunk 3]
```diff
@@ -1712,6 +1712,8 @@ BOOST_AUTO_TEST_CASE(ThreeSolversParallel)
   runTestThreeSolvers(config, expectedCallsOfAdvance, context);
 }
 
+// @todo mapping seems to be not set up properly?
+#if 0
```
Comment (uekerman) [inline src/precice/tests/SerialTests.cpp:1716]: Dangerous strategy. Why not keep the test failing for the moment?

File: src/time/Waveform.cpp [hunk 4]
```diff
@@ -58,10 +59,12 @@ Eigen::VectorXd Waveform::sample(double normalizedDt)
 void Waveform::moveToNextWindow()
 {
   PRECICE_ASSERT(_storageIsInitialized);
-  auto initialGuess = extrapolate();
-  utils::shiftSetFirst(this->_timeWindowsStorage, initialGuess); // archive old samples and store initial guess
-  if (_numberOfStoredSamples < sizeOfSampleStorage()) {          // together with the initial guess the number of stored samples increases
-    _numberOfStoredSamples++;
+  if (_extrapolationOrder != Time::UNDEFINED_EXTRAPOLATION_ORDER) { // @todo: we should define the extrapolation data for all data (if the waveform is owned by the mesh::Data), another argument for configuring extrapolation order there.
```
Comment (uekerman) [inline src/time/Waveform.cpp:62]: Currently, `_extrapolationOrder` is a property of a coupling scheme. Not all data is extrapolated (not data sent from first to second in a serial coupling scheme). And different coupling schemes can have different orders. Do you have the latter on the radar?

File: src/mesh/Data.hpp (hunk 2 above)
Comment (BenjaminRodenberg) [reply src/mesh/Data.hpp:28]: This is work-in-progress and will definitely vanish when we have a proper configuration.

**Pull request merged**